src/pudl/extract/csv.py

```diff
@@ -0,0 +1,153 @@
+"""Extractor for CSV data."""
+from csv import DictReader
+from functools import lru_cache
+from importlib import resources
+from pathlib import Path
+from zipfile import ZipFile
+
+import pandas as pd
+import sqlalchemy as sa
+
+from pudl import logging_helpers
+from pudl.workspace.datastore import Datastore
+
+logger = logging_helpers.get_logger(__name__)
+
+
+class CsvReader:
+    """Wrapper to provide standardized access to CSV files."""
+
+    def __init__(self, datastore: Datastore, dataset: str):
+        """Create a new instance of CsvReader.
+
+        This can be used for retrieving data from CSV files.
+
+        Args:
+            datastore: provides access to raw files on disk.
+            dataset: name of the dataset (e.g. eia176), this is used to load metadata
+                from package_data/{dataset} subdirectory.
+        """
+        self.datastore = datastore
+        self.dataset = dataset
+        self._table_file_map = {}
+        for row in self._open_csv_resource("table_file_map.csv"):
+            self._table_file_map[row["table"]] = row["filename"]
+
+    def _open_csv_resource(self, base_filename: str) -> DictReader:
+        """Open the given resource file as :class:`csv.DictReader`."""
+        csv_path = resources.files(f"pudl.package_data.{self.dataset}") / base_filename
+        return DictReader(csv_path.open())
+
+    def get_table_names(self) -> list[str]:
+        """Returns list of tables that this datastore provides access to."""
+        return list(self._table_file_map)
+
+    @lru_cache
+    def _cache_zipfile(self) -> ZipFile:
+        """Returns a ZipFile instance corresponding to the dataset."""
+        return self.datastore.get_zipfile_resource(self.dataset)
+
+    def read(self, filename: str) -> pd.DataFrame:
+        """Read the data from the CSV source and return as dataframes."""
+        logger.info(f"Extracting {filename} from CSV into pandas DataFrame.")
+        zipfile = self._cache_zipfile()
+        with zipfile.open(filename) as f:
+            # TODO: Define encoding
+            df = pd.read_csv(f)
+        return df
+
+
+class CsvExtractor:
```

Comment:
This implementation was based on `FercDbfExtractor`. So far I've been able to slim it down some compare to that class.

Reply:
Ah, here's something confusing about our system:

* FERC DBF data gets converted to tables `ferc{form_number}.sqlite` outside of the normal ETL process - I don't know why, actually. FERC XBRL gets converted to a separate table also, that's because we don't know what the schema is going to be until we've read in the dang XML file.
* EIA data gets extracted to `raw_eia176__....` pickled dataframes in the Dagster asset storage directory - see `src/pudl/extract/eia860.py` for how that gets executed. They then feed into `_core_eia176__...` assets which also are pickled, which *finally* feed into `output_eia176__...` assets which get written to `pudl.sqlite` ( see `src/pudl/transform/eia860.py` and `src/pudl/output/eia860.py` for details).

Long story short, we don't need to worry about where the data is getting persisted on disk at this juncture, we just need to pull out a bunch of dataframes. I think if you want to test the `CsvExtractor` through testing the `Eia176CsvExtractor`, you'd want to make a dagster asset in ``extract/eia176.py` that looks something like....


```python
@asset
def raw_eia176__company(context):
    extractor = Eia176CsvExtractor(context.resource.datastore)
    # ... get the company table as dataframe
    # return that dataframe
```

And then you should be able to just test `raw_eia176_company` like a normal function, though you would need to mock out the `context.resource.datastore`.

Reply:
That's helpful, thanks. I'll work on an iteration along those lines and get something surfaced by Monday.

Reply:
Sweet! Just an FYI - a bunch of us are gonna be at a conference Monday and Tuesday. But ping on Slack if you're ready and I'll pull some time to take a look!